### PR TITLE
#446 DatagramServer.join()/leave() return Promise<Boolean>; fix UdpServe...

### PR DESCRIPTION
**Pull request by lukejackson**
See https://github.com/reactor/reactor/issues/446. This doesn't tackle the problem with void futures, but works around them by making `DatagramServer`.`join()`/`leave()` return `Promise<Boolean>`, consistent with the other asynchronous methods on this class.

At the same time I have:

* Fixed and reenabled the UdpServerTests 
* Removed duplicate channel configuration - setting Netty bootstrap options is sufficient
* Add ProtocolFamily server socket option - this is to get the multicast test to pass on a machine with IPv6 support. The Multicast test explicitly uses an IPv4 multicast address, so needs a IPv4 DatagramChannel.

### CODE REVIEW

File: reactor-net/src/main/java/reactor/io/net/impl/netty/tcp/NettyTcpServer.java [hunk 1]
```diff
@@ -99,22 +98,16 @@ protected NettyTcpServer(@Nonnull Environment env,
...
 				.childHandler(new ChannelInitializer<SocketChannel>() {
 					@Override
 					public void initChannel(final SocketChannel ch) throws Exception {
-						SocketChannelConfig config = ch.config();
-						config.setReceiveBufferSize(options.rcvbuf());
-						config.setSendBufferSize(options.sndbuf());
-						config.setKeepAlive(options.keepAlive());
-						config.setReuseAddress(options.reuseAddr());
-						config.setSoLinger(options.linger());
-						config.setTcpNoDelay(options.tcpNoDelay());
```
Comment (lukejackson) [inline reactor-net/src/main/java/reactor/io/net/impl/netty/tcp/NettyTcpServer.java]: no need to set these again. options will have already been set by Netty. Adding the missing options above.

File: reactor-net/src/main/java/reactor/io/net/impl/netty/udp/NettyDatagramServer.java [hunk 2]
```diff
@@ -88,35 +88,28 @@ public NettyDatagramServer(@Nonnull Environment env,
...
 				.group(ioGroup)
 				.option(ChannelOption.SO_RCVBUF, options.rcvbuf())
 				.option(ChannelOption.SO_SNDBUF, options.sndbuf())
 				.option(ChannelOption.SO_REUSEADDR, options.reuseAddr())
+				.option(ChannelOption.AUTO_READ, false)
+				.option(ChannelOption.CONNECT_TIMEOUT_MILLIS, options.timeout())
 				.channelFactory(new ChannelFactory<Channel>() {
+									@Override
+									public Channel newChannel() {
+										return new NioDatagramChannel(family);
```
Comment (lukejackson) [inline reactor-net/src/main/java/reactor/io/net/impl/netty/udp/NettyDatagramServer.java]: allow IPv4 / IPv6 to be specified

File: reactor-net/src/main/java/reactor/io/net/impl/netty/udp/NettyDatagramServer.java [hunk 3]
```diff
@@ -88,35 +88,28 @@ public NettyDatagramServer(@Nonnull Environment env,
...
-						final NioDatagramChannel ch = new NioDatagramChannel();
-						DatagramChannelConfig config = ch.config();
-						config.setReceiveBufferSize(options.rcvbuf());
-						config.setSendBufferSize(options.sndbuf());
-						config.setReuseAddress(options.reuseAddr());
-						config.setAutoRead(false);
-
-						if (null != multicastInterface) {
-							config.setNetworkInterface(multicastInterface);
-						}
```
Comment (lukejackson) [inline reactor-net/src/main/java/reactor/io/net/impl/netty/udp/NettyDatagramServer.java]: all this either already set above or below

File: reactor-net/src/test/java/reactor/io/net/udp/UdpServerTests.java [hunk 4]
```diff
@@ -73,7 +72,7 @@ public void accept(byte[] bytes) {
 			try {
 				DatagramChannel udp = DatagramChannel.open();
 				udp.configureBlocking(true);
-				udp.connect(new InetSocketAddress(port));
+				udp.connect(new InetSocketAddress(InetAddress.getLocalHost(), port));
```
Comment (lukejackson) [inline reactor-net/src/test/java/reactor/io/net/udp/UdpServerTests.java]: can't connect to 0.0.0.0 - test was failing here

File: reactor-net/src/test/java/reactor/io/net/udp/UdpServerTests.java [hunk 5]
```diff
@@ -127,10 +125,11 @@ public void accept(byte[] bytes) {
 						latch.countDown();
 					}
 				}
-			});
+			}));
+
+			server.start().onSuccess(b -> server.join(multicastGroup, multicastIface)).await();
```
Comment (lukejackson) [inline reactor-net/src/test/java/reactor/io/net/udp/UdpServerTests.java]: such chaining now works that the promise is not completed will null

**Comment by jbrisbin** [comment]
Thanks for the PR!

**Comment by jbrisbin** [comment]
@lukejackson This seems to work okay on the CI server, which is linux. But my Mac doesn't like it (OS X Yosemite). I keep getting:

    Caused by: java.io.IOException: Network interface not configured for IPv4
	    at sun.nio.ch.DatagramChannelImpl.setOption(DatagramChannelImpl.java:226) ~[na:1.8.0]
	    at sun.nio.ch.DatagramChannelImpl.setOption(DatagramChannelImpl.java:41) ~[na:1.8.0]

**Comment by smaldini** [comment]
Sorry I have to comment it I needed to pass a fix for take() and takeWhile(), feel free to continue.

**Comment by smaldini** [comment]
Don't pay attention to commit message :)

**Comment by lukejackson** [comment]
Raised https://github.com/reactor/reactor/pull/448 to check the interface has an IPv4 address, and also use the loopback interface where possible. On my Mac at home I see the loopback interface is multicast enabled.

More details in the PR.

**Pull request merged**